src/Logary/ValueModule.fs

```diff
@@ -0,0 +1,252 @@
+namespace Logary
+
+open System
+open Microsoft.FSharp.Reflection // TODO: move to type-shape
+open System.Collections.Generic // TODO: move to type-shape
+open Logary.Internals.Aether
+open Logary
+open Logary.Internals // TODO: move to type-shape, for Map.ofObject
+
+[<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
+module Value =
+
+  module Optic = 
+
+    (* Epimorphisms *)
+
+    let internal String__ =
+      (function | String x -> Some x
+                | _ -> None), String
+
+    let internal Bool__ =
+      (function | Bool x -> Some x
+                | _ -> None), Bool
+
+    let internal Float__ =
+      (function | Float x -> Some x
+                | _ -> None), Float
+
+    let internal Int64__ =
+      (function | Int64 x -> Some x
+                | _ -> None), Int64
+
+    let internal BigInt__ =
+      (function | BigInt x -> Some x
+                | _ -> None), BigInt
+
+    let internal Binary__ =
+      (function | Binary (bs, ct) -> Some (bs, ct)
+                | _ -> None), Binary
+
+    let internal Fraction__ =
+      (function | Fraction (n, d) -> Some (n, d)
+                | _ -> None), Fraction
+
+    let internal Array__ =
+      (function | Array x -> Some x
+                | _ -> None), Array
+
+    let internal Object__ =
+      (function | Object x -> Some x
+                | _ -> None), Object
+
+    (* Prisms *)
+
+    let String_ =
+      Prism.ofEpimorphism String__
+
+    let Bool_ =
+      Prism.ofEpimorphism Bool__
+
+    let Float_ =
+      Prism.ofEpimorphism Float__
+
+    let Int64_ =
+      Prism.ofEpimorphism Int64__
+
+    let BigInt_ =
+      Prism.ofEpimorphism BigInt__
+
+    let Binary_ =
+      Prism.ofEpimorphism Binary__
+
+    let Fraction_ =
+      Prism.ofEpimorphism Fraction__
+
+    let Array_ =
+      Prism.ofEpimorphism Array__
+
+    let Object_ =
+      Prism.ofEpimorphism Object__
+
+  // TODO: inline, 'a -> Value
+  let rec internal exceptionToStringValueMap (valueOfObject : obj -> Value) (e : exn) =
+    let fields =
+      [ yield "type", String (e.GetType ()).FullName
+        yield "message", String e.Message
+        if e.TargetSite <> null then
+          yield "targetSite", String (e.TargetSite.ToString ())
+        if e.StackTrace <> null then
+          yield "stackTrace", String e.StackTrace
+        if e.Source <> null then
+          yield "source", String e.Source
+        if e.HelpLink <> null then
+          yield "helpLink", String e.HelpLink
+        if e.HResult <> 0 then
+          yield "hResult", Int64 (int64 e.HResult)
+        if e.Data <> null && e.Data.Count > 0 then
+          yield "data", valueOfObject e.Data ]
+
+    HashMap.ofList <|
+      if e.InnerException <> null then
+        ("inner", Object <| exceptionToStringValueMap valueOfObject e.InnerException) :: fields
+      else
+        fields
+
+  // TODO: move out of this file,
+  // TODO: use TypeShape,
```

Comment:
It caches the boxing + reflection and makes it possible to create values configurable: by default we should call `ToString()` on them, but allow the user to add more `Visitor`s to the configuration.